### ISSUE TITLE
Implicitly add non-exisitng incident nodes upon edge insertion

### ISSUE THREAD

**Issue by JunpengCode** [networkit/networkit#419]
The following code is wrong: segmentation fault,
```python
import networkit as nk
if __name__ == '__main__':
    G= nk.graph.Graph()
    G.addEdge(0,1)
    print(G.numberOfNodes())
```
It is due to that the node index is not exist.  The following code is correct,
```python
import networkit as nk
if __name__ == '__main__':
    G= nk.graph.Graph()
    G.addNode()
    G.addNode()
    G.addNode()
    G.addEdge(0,1)
    print(G.numberOfNodes())
```

I consider that the function may be added into the networkit, In the networkx library, These are added automatically when edges are added.

**Comment by manpen** [comment]
I would vote against doing this implicitly. IMHO adding edges to non-existing nodes should be illegal as you're using unallocated memory. You would not expect the following code to work either:

```Python
x = [0]
x[1] = 1
```

Having an error message for the Python interface (which I can add as soon as we agree on this matter) should help to uncover potential logic bugs. And if you really want to add the nodes you can do it manually. A compromise would be flag that automatically increases the set of nodes, e.g.,

```Python
G = nk.Graph()
G.addEdge(0, 1, addNodes = True)
```

**Comment by JunpengCode** [comment]
I think that  adding the flag that automatically increases the set of nodes is reasonable.

**Comment by avdgrinten** [comment]
I agree with Manuel. The number of implicit functionality in the C++ core should be kept to a minimum. If this feature is *really* desired, we could add a helper function in addition to `addEdge`.

**Comment by JunpengCode** [comment]
I consider that  adding this feature makes networkit more better. For example, the file format is following:
1 2 3 4 5
2 3 4 5 6
3 4 5 6 7

I may be selected the col 1,3, and 5 as node1, node2 and weight.  This avoids adding more read functions.  And it is also more flexible.

**Comment by JunpengCode** [comment]
For example:

```python
def readUndirectedGraph(file):
    G = nk.graph.Graph()
    f = open(file)
    try:
        for line in f.readlines():
            cell = line.split()
            G.add_edge(cell[0],cell[1])
    except:
        print("Unexpected error:", sys.exc_info())
        input('press enter key to exit')
    return G
```

**Comment by JunpengCode** [comment]
```python
def readUndirectedGraph(file):
    G = nk.graph.Graph()
    f = open(file)
    try:
        for line in f.readlines():
            cell = line.split()
            G.add_edge(cell[0],cell[1])
    except:
        print("Unexpected error:", sys.exc_info())
        input('press enter key to exit')
    return G
```

**Comment by manpen** [comment]
The problem with adding nodes is that it's potentially expensive -- in the worst case (directed, weighted graphs with edge ids) it can trigger 6 vectors to be reallocated and copied. Hence, from the perspective of the C++ core it's almost always better to first figure out how many nodes are needed, and then to create a Graph of the correct size.

This trade-off, however, maybe quite different for the Python interface. So I can see that this feature may not only be about convenience but performance. So, if we want to have it:
- Should it be implemented as a flag (see my example) or extra function `addEdgeWithNodes` (I do not like the name)
- Should it restore previously deleted nodes (I'd say so)
- Should it be available in the C++ core? (I'd say so, despite the fear of encouraging the anti-pattern mentioned above)

@JunpengCode Would you like to implement it?

**Comment by JunpengCode** [comment]
Or, Is the function convert to networkx op implicity? For C++ implementations, it does have a big performance problem.

**Comment by JunpengCode** [comment]
Or, Is the function convert to networkx op implicity? For C++ implementations, it does have a big performance problem. I guess.

**Comment by manpen** [comment]
Can you please elaborate on this?

**Comment by JunpengCode** [comment]
(1)Adding the "addNodes" flag;
(2)Using the nk2nx convert the op;
(3)call the c++ function to storage the graph.

I guess.

**Comment by JunpengCode** [comment]
delay the write op until graph creation using addEdges

**Comment by manpen** [comment]
A fix is provided in #433
